File: sdks/python/hub_router.py
"""
hub-router SDK — single-file Python client.

Copy this file into your project. No pip install required. Requires Python 3.9+.

Usage::

    from hub_router import OnlineClient, LocalClient

    # Online side — submit and wait for results
    client = OnlineClient("https://hub.example.com", api_key="secret")
    result = await client.do({"query": "hello"})

    # Forward an incoming FastAPI/Starlette request verbatim
    result = await client.do_request(request)

    # Local side — process requests from the queue
    local = LocalClient("https://hub.example.com", api_key="local-secret")
    async def process(req):
        return Result(request_id=req.id, payload=await my_model(req.payload))
    await local.run(process)
"""
from __future__ import annotations

import asyncio
import json
import urllib.error
import urllib.request
from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Any, Awaitable, Callable, Optional

# ─── Exceptions ───────────────────────────────────────────────────────────────


class HubRouterError(Exception):
    """Base exception for all hub-router errors."""


class QueueFullError(HubRouterError):
    """Raised when the middleware queue is at capacity."""


class RequestNotFoundError(HubRouterError):
    """Raised when a request ID is not found (expired or never existed)."""


class ResultTimeoutError(HubRouterError):
    """Raised when wait_result exhausts all retries without receiving a result."""


# ─── Models ───────────────────────────────────────────────────────────────────


@dataclass
class QueuedRequest:
    """A pending request pulled by the local server."""

    id: str
    payload: Any
    headers: dict[str, str] = field(default_factory=dict)
    enqueued_at: Optional[datetime] = None
    expires_at: Optional[datetime] = None

    @classmethod
    def from_dict(cls, data: dict) -> "QueuedRequest":
        return cls(
            id=data["id"],
            payload=data["payload"],
            headers=data.get("headers") or {},
            enqueued_at=_parse_dt(data.get("enqueued_at")),
            expires_at=_parse_dt(data.get("expires_at")),
        )


@dataclass
class Result:
    """A processed result pushed back by the local server."""

    request_id: str
    payload: Any = None
    status_code: int = 200
    error: str = ""
    completed_at: Optional[datetime] = None

    def to_dict(self) -> dict:
        return {
            "request_id": self.request_id,
            "payload": self.payload,
            "status_code": self.status_code,
            "error": self.error,
            "completed_at": self.completed_at.isoformat() if self.completed_at else None,
        }

    @classmethod
    def from_dict(cls, data: dict) -> "Result":
        return cls(
            request_id=data["request_id"],
            payload=data.get("payload"),
            status_code=data.get("status_code", 200),
            error=data.get("error", ""),
            completed_at=_parse_dt(data.get("completed_at")),
        )


def _parse_dt(value: Optional[str]) -> Optional[datetime]:
    if not value:
        return None
    try:
        return datetime.fromisoformat(value.replace("Z", "+00:00"))
    except (ValueError, AttributeError):
        return None


# ─── HTTP helper (stdlib, no external deps) ───────────────────────────────────

_HOP_BY_HOP = frozenset(
    {
        "content-length",
        "host",
        "transfer-encoding",
        "connection",
        "keep-alive",
        "proxy-authenticate",
        "proxy-authorization",
        "te",
        "trailers",
        "upgrade",
    }
)


async def _http(
    method: str,
    url: str,
    *,
    data: Optional[bytes] = None,
    headers: Optional[dict[str, str]] = None,
) -> tuple[int, bytes]:
    """Non-blocking HTTP request using stdlib urllib wrapped in a thread."""

    def _blocking() -> tuple[int, bytes]:
        req = urllib.request.Request(
            url,
            data=data,
            headers=headers or {},
            method=method,
        )
        try:
            with urllib.request.urlopen(req) as resp:
                return resp.status, resp.read()
        except urllib.error.HTTPError as exc:
            return exc.code, exc.read()

    return await asyncio.to_thread(_blocking)


# ─── Type alias ──────────────────────────────────────────────────────────────

ProcessorFunc = Callable[["QueuedRequest"], Awaitable["Result"]]


# ─── OnlineClient ─────────────────────────────────────────────────────────────


class OnlineClient:
    """
    Async client for the online (web app) side of hub-router.

    All methods are non-blocking — they never block the event loop.
    Compatible with FastAPI, Starlette, Django async, or any asyncio app.

    Usage::

        client = OnlineClient("https://hub.example.com", api_key="secret")

        # Submit a payload and wait for the result
        result = await client.do({"query": "hello"})
        print(result.payload)

        # Forward an incoming request verbatim (FastAPI example)
        @app.post("/infer")
        async def infer(request: Request):
            result = await client.do_request(request)
            return result.payload
    """

    def __init__(
        self,
        base_url: str,
        api_key: str,
        *,
        long_poll_timeout: float = 30.0,
        max_retries: int = 10,
    ) -> None:
        self._base_url = base_url.rstrip("/")
        self._api_key = api_key
        self._long_poll_timeout = long_poll_timeout
        self._max_retries = max_retries

    async def submit(
        self,
        payload: Any,
        headers: Optional[dict[str, str]] = None,
    ) -> str:
        """
        Enqueue a request and return its correlation ID.

        :raises QueueFullError: If the middleware queue is at capacity.
        """
        body = json.dumps({"payload": payload, "headers": headers or {}}).encode()
        status, raw = await _http(
            "POST",
            f"{self._base_url}/request",
            data=body,
            headers={
                "Content-Type": "application/json",
                "X-Online-API-Key": self._api_key,
            },
        )
        if status == 503:
            raise QueueFullError(json.loads(raw).get("message", "queue full"))
        if status != 202:
            raise HubRouterError(f"unexpected status {status}: {raw.decode()}")
        return json.loads(raw)["id"]

    async def wait_result(
        self,
        request_id: str,
        *,
        timeout: Optional[float] = None,
    ) -> Result:
        """
        Long-poll until the result for *request_id* is ready.

        :param request_id: The ID returned by :meth:`submit`.
        :param timeout: Per-poll timeout in seconds (defaults to ``long_poll_timeout``).
        :raises RequestNotFoundError: If the ID is unknown or expired.
        :raises ResultTimeoutError: If ``max_retries`` 204 responses are received.
        """
        t = timeout or self._long_poll_timeout
        url = f"{self._base_url}/result/{request_id}?timeout={t}s"
        hdrs = {"X-Online-API-Key": self._api_key}

        for _ in range(self._max_retries + 1):
            status, raw = await _http("GET", url, headers=hdrs)
            if status == 200:
                return Result.from_dict(json.loads(raw))
            if status == 204:
                continue
            if status == 404:
                raise RequestNotFoundError(
                    f"request ID {request_id!r} not found (may have expired)"
                )
            raise HubRouterError(f"unexpected status {status}: {raw.decode()}")

        raise ResultTimeoutError(
            f"result not available after {self._max_retries} retries "
            f"for request {request_id!r}"
        )

    async def do_sync(
        self,
        payload: Any,
        headers: Optional[dict[str, str]] = None,
        *,
        timeout: Optional[float] = None,
    ) -> tuple[Optional[Result], Optional[str]]:
        """
        Single-request fast path using POST /request/sync.

        Returns ``(Result, None)`` if the local server responds before the timeout,
        or ``(None, request_id)`` on slow path — continue with :meth:`wait_result`.

        :raises QueueFullError: If the queue is at capacity.
        """
        t = timeout or self._long_poll_timeout
        body = json.dumps({"payload": payload, "headers": headers or {}}).encode()
        status, raw = await _http(
            "POST",
            f"{self._base_url}/request/sync?timeout={t}s",
            data=body,
            headers={
                "Content-Type": "application/json",
                "X-Online-API-Key": self._api_key,
            },
        )
        if status == 503:
            raise QueueFullError(json.loads(raw).get("message", "queue full"))
        if status == 200:
            return Result.from_dict(json.loads(raw)), None
        if status == 202:
            return None, json.loads(raw)["id"]
        raise HubRouterError(f"unexpected status {status}: {raw.decode()}")

    async def do(
        self,
        payload: Any,
        headers: Optional[dict[str, str]] = None,
    ) -> Result:
        """
        Submit a request and wait for its result.

        Tries the sync fast path first (single round-trip when local server is fast).
        Falls back to async polling transparently when local server is slow.
        """
        result, request_id = await self.do_sync(payload, headers)
        if result is not None:
            return result
        return await self.wait_result(request_id)  # type: ignore[arg-type]

    async def do_request(
        self,
        request: Any,
        *,
        timeout: Optional[float] = None,
    ) -> Result:
        """
        Forward an incoming HTTP request verbatim to hub-router.

        The original headers and body are passed through to the local server
        unchanged. Hop-by-hop headers (Content-Length, Host, etc.) are excluded.

        Accepts any request object with:
        - ``.headers`` — dict-like mapping of header name → value
        - ``.body()`` — async method (or ``.body`` attribute) returning bytes

        Compatible with FastAPI and Starlette ``Request`` objects::

            @app.post("/infer")
            async def infer(request: Request):
                result = await client.do_request(request)
                return result.payload
        """
        # Extract headers, skipping hop-by-hop
        fwd_headers: dict[str, str] = {}
        raw_headers = getattr(request, "headers", {})
        items = raw_headers.items() if hasattr(raw_headers, "items") else raw_headers
        for k, v in items:
            if k.lower() not in _HOP_BY_HOP:
                fwd_headers[k] = v

        # Read body — support async .body() coroutine or sync .body attribute
        body_attr = getattr(request, "body", None)
        if callable(body_attr):
            raw_body = await body_attr()
        elif body_attr is not None:
            raw_body = body_attr
        else:
            raw_body = b""

        payload = (
            raw_body.decode("utf-8", errors="replace")
            if isinstance(raw_body, bytes)
            else str(raw_body or "")
        )
        return await self.do(payload, fwd_headers)


# ─── LocalClient ─────────────────────────────────────────────────────────────


class LocalClient:
    """
    Async client for the local processing server side of hub-router.

    All methods are non-blocking — they never block the event loop.

    Usage::

        async def process(req: QueuedRequest) -> Result:
            answer = await my_model.infer(req.payload)
            return Result(request_id=req.id, payload=answer)

        client = LocalClient("https://hub.example.com", api_key="local-secret",
                             batch_size=20, workers=5)
        # Run alongside other async tasks:
        task = asyncio.create_task(client.run(process))
        # Or block:
        await client.run(process)
    """

    def __init__(
        self,
        base_url: str,
        api_key: str,
        *,
        batch_size: int = 10,
        poll_interval: float = 1.0,
        workers: int = 1,
    ) -> None:
        self._base_url = base_url.rstrip("/")
        self._api_key = api_key
        self._batch_size = batch_size
        self._poll_interval = poll_interval
        self._workers = workers
        self._stop_event = asyncio.Event()

    def stop(self) -> None:
        """Signal the poll loop to stop after the current batch finishes."""
        self._stop_event.set()

    async def run(self, processor: ProcessorFunc) -> None:
        """
        Start the poll loop. Blocks until :meth:`stop` is called or task is cancelled.

        Dispatches each request as a separate asyncio task, bounded by the
        ``workers`` semaphore to prevent concurrency explosion.
        """
        sem = asyncio.Semaphore(self._workers)
        pending_tasks: set[asyncio.Task] = set()

        while not self._stop_event.is_set():
            requests = await self.pull_batch()

            if not requests:
                try:
                    await asyncio.wait_for(
                        asyncio.shield(self._stop_event.wait()),
                        timeout=self._poll_interval,
                    )
                    break
                except asyncio.TimeoutError:
                    continue

            for req in requests:
                task = asyncio.create_task(
                    self._process_and_push(sem, processor, req)
                )
                pending_tasks.add(task)
                task.add_done_callback(pending_tasks.discard)

        if pending_tasks:
            await asyncio.gather(*pending_tasks, return_exceptions=True)

    async def pull_batch(self) -> list[QueuedRequest]:
        """Fetch up to ``batch_size`` pending requests. Never blocks."""
        status, raw = await _http(
            "GET",
            f"{self._base_url}/queue/pull?batch={self._batch_size}",
            headers={"X-Local-API-Key": self._api_key},
        )
        if status != 200:
            raise HubRouterError(f"unexpected status {status}: {raw.decode()}")
        data = json.loads(raw)
        return [QueuedRequest.from_dict(r) for r in data.get("requests", [])]

    async def push_result(self, result: Result) -> None:
        """Push a completed result to hub-router."""
        if result.completed_at is None:
            result.completed_at = datetime.now(timezone.utc)
        body = json.dumps(result.to_dict()).encode()
        status, raw = await _http(
            "POST",
            f"{self._base_url}/queue/result",
            data=body,
            headers={
                "Content-Type": "application/json",
                "X-Local-API-Key": self._api_key,
            },
        )
        if status == 404:
            raise RequestNotFoundError(
                f"request ID {result.request_id!r} not found (may have expired)"
            )
        if status != 204:
            raise HubRouterError(f"unexpected status {status}: {raw.decode()}")

    async def _process_and_push(
        self,
        sem: asyncio.Semaphore,
        processor: ProcessorFunc,
        req: QueuedRequest,
    ) -> None:
        async with sem:
            try:
                result = await processor(req)
            except Exception as exc:  # noqa: BLE001
                result = Result(
                    request_id=req.id,
                    payload={"error": str(exc)},
                    status_code=500,
                    error=str(exc),
                )
            try:
                await self.push_result(result)
            except HubRouterError:
                pass  # best-effort; request may have expired

File: sdks/node/src/hub-router.ts
/**
 * hub-router SDK — single-file TypeScript client.
 *
 * Copy this file into your project. No npm install required.
 * Requires Node.js 18+ (native fetch) or any environment with the Web Fetch API.
 *
 * Usage:
 *   import { OnlineClient, LocalClient } from "./hub-router.js";
 *
 *   const client = new OnlineClient("https://hub.example.com", "api-key");
 *   const result = await client.do({ query: "hello" });
 *
 *   // Forward an incoming HTTP request verbatim:
 *   const result = await client.doRequest(req);
 */

// ─── Types ────────────────────────────────────────────────────────────────────

/** A pending request pulled by the local server. */
export interface QueuedRequest {
  id: string;
  payload: unknown;
  headers: Record<string, string>;
  enqueued_at: string;
  expires_at: string;
}

/** A processed result pushed back by the local server. */
export interface Result {
  request_id: string;
  payload: unknown;
  status_code: number;
  error?: string;
  completed_at?: string;
}

/** Response from POST /request or POST /request/sync (202). */
export interface SubmitResponse {
  id: string;
  estimated_wait_ms: number;
}

/** Response from GET /queue/pull. */
export interface PullResponse {
  requests: QueuedRequest[];
  count: number;
}

/** Options for OnlineClient. */
export interface OnlineClientOptions {
  /** Long-poll timeout in milliseconds (default: 30_000). */
  longPollTimeout?: number;
  /** Max number of 204 retries before giving up (default: 10). */
  maxRetries?: number;
}

/** Options for LocalClient. */
export interface LocalClientOptions {
  /** Requests per poll cycle (default: 10). */
  batchSize?: number;
  /** Sleep in ms when queue is empty (default: 1000). */
  pollInterval?: number;
  /** Max concurrent processor calls (default: 1). */
  workers?: number;
}

/** Processor callback: receives a request, returns a result. */
export type ProcessorFn = (req: QueuedRequest) => Promise<Result>;

// ─── Errors ──────────────────────────────────────────────────────────────────

export class HubRouterError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "HubRouterError";
  }
}

export class QueueFullError extends HubRouterError {
  constructor(message = "queue is full") {
    super(message);
    this.name = "QueueFullError";
  }
}

export class RequestNotFoundError extends HubRouterError {
  constructor(requestId: string) {
    super(`request ID '${requestId}' not found (may have expired)`);
    this.name = "RequestNotFoundError";
  }
}

export class ResultTimeoutError extends HubRouterError {
  constructor(requestId: string, retries: number) {
    super(
      `result not available after ${retries} retries for request '${requestId}'`
    );
    this.name = "ResultTimeoutError";
  }
}

// ─── OnlineClient ─────────────────────────────────────────────────────────────

const HOP_BY_HOP = new Set([
  "content-length",
  "host",
  "transfer-encoding",
  "connection",
  "keep-alive",
  "proxy-authenticate",
  "proxy-authorization",
  "te",
  "trailers",
  "upgrade",
]);

const DEFAULT_LONG_POLL_TIMEOUT_MS = 30_000;
const DEFAULT_MAX_RETRIES = 10;

/**
 * OnlineClient — submit requests and wait for results from hub-router.
 * Fully async, non-blocking. Works with Express, Fastify, NestJS, etc.
 *
 * ```ts
 * const client = new OnlineClient("https://hub.example.com", "api-key");
 *
 * // Simple payload
 * const result = await client.do({ query: "hello" });
 *
 * // Forward an incoming HTTP request verbatim (headers + body pass-through)
 * app.post("/api/infer", async (req, res) => {
 *   const result = await client.doRequest(req);
 *   res.json(result.payload);
 * });
 * ```
 */
export class OnlineClient {
  private readonly baseUrl: string;
  private readonly apiKey: string;
  private readonly longPollTimeoutMs: number;
  private readonly maxRetries: number;

  /**
   * @param baseUrl  hub-router base URL, e.g. "https://hub.example.com"
   * @param apiKey   X-Online-API-Key value
   * @param options  Optional overrides
   */
  constructor(
    baseUrl: string,
    apiKey: string,
    options: OnlineClientOptions = {}
  ) {
    this.baseUrl = baseUrl.replace(/\/$/, "");
    this.apiKey = apiKey;
    this.longPollTimeoutMs =
      options.longPollTimeout ?? DEFAULT_LONG_POLL_TIMEOUT_MS;
    this.maxRetries = options.maxRetries ?? DEFAULT_MAX_RETRIES;
  }

  /**
   * Enqueue a request and return its correlation ID.
   * @throws {QueueFullError} if the queue is at capacity.
   */
  async submit(
    payload: unknown,
    headers: Record<string, string> = {}
  ): Promise<string> {
    const res = await fetch(`${this.baseUrl}/request`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "X-Online-API-Key": this.apiKey,
      },
      body: JSON.stringify({ payload, headers }),
    });

    if (res.status === 503) {
      const data = (await res.json()) as { message?: string };
      throw new QueueFullError(data.message);
    }
    if (res.status !== 202) {
      throw new HubRouterError(
        `unexpected status ${res.status}: ${await res.text()}`
      );
    }

    const data = (await res.json()) as SubmitResponse;
    return data.id;
  }

  /**
   * Long-poll until the result for *requestId* is ready.
   * Retries automatically on 204 (server timeout) up to `maxRetries` times.
   *
   * @throws {RequestNotFoundError} if the ID is unknown or expired.
   * @throws {ResultTimeoutError}   if `maxRetries` exhausted.
   */
  async waitResult(
    requestId: string,
    options?: { timeout?: number }
  ): Promise<Result> {
    const timeoutMs = options?.timeout ?? this.longPollTimeoutMs;
    const timeoutSec = (timeoutMs / 1000).toFixed(1);
    const url = `${this.baseUrl}/result/${encodeURIComponent(requestId)}?timeout=${timeoutSec}s`;

    for (let attempt = 0; attempt <= this.maxRetries; attempt++) {
      const res = await fetch(url, {
        headers: { "X-Online-API-Key": this.apiKey },
        signal: AbortSignal.timeout(timeoutMs + 10_000),
      });

      if (res.status === 200) return (await res.json()) as Result;
      if (res.status === 204) continue;
      if (res.status === 404) throw new RequestNotFoundError(requestId);
      throw new HubRouterError(
        `unexpected status ${res.status}: ${await res.text()}`
      );
    }

    throw new ResultTimeoutError(requestId, this.maxRetries);
  }

  /**
   * Single-request fast path using POST /request/sync.
   *
   * Returns `{ result }` when the local server responds before the timeout,
   * or `{ requestId }` when it is slow — caller should then call {@link waitResult}.
   *
   * @throws {QueueFullError} if the queue is at capacity.
   */
  async doSync(
    payload: unknown,
    headers: Record<string, string> = {},
    options?: { timeout?: number }
  ): Promise<{ result: Result } | { requestId: string }> {
    const timeoutMs = options?.timeout ?? this.longPollTimeoutMs;
    const timeoutSec = (timeoutMs / 1000).toFixed(1);
    const url = `${this.baseUrl}/request/sync?timeout=${timeoutSec}s`;

    const res = await fetch(url, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "X-Online-API-Key": this.apiKey,
      },
      body: JSON.stringify({ payload, headers }),
      signal: AbortSignal.timeout(timeoutMs + 10_000),
    });

    if (res.status === 200) return { result: (await res.json()) as Result };
    if (res.status === 202) {
      const data = (await res.json()) as { id: string };
      return { requestId: data.id };
    }
    if (res.status === 503) {
      const data = (await res.json()) as { message?: string };
      throw new QueueFullError(data.message);
    }
    throw new HubRouterError(
      `unexpected status ${res.status}: ${await res.text()}`
    );
  }

  /**
   * Submit a request and wait for its result.
   * Tries the sync fast path first; falls back to async polling if the local server is slow.
   */
  async do(
    payload: unknown,
    headers: Record<string, string> = {}
  ): Promise<Result> {
    const response = await this.doSync(payload, headers);
    if ("result" in response) return response.result;
    return this.waitResult(response.requestId);
  }

  /**
   * Forward an incoming HTTP request verbatim to hub-router.
   *
   * The original headers and body are passed through to the local server unchanged.
   * Hop-by-hop headers (Content-Length, Host, Transfer-Encoding, etc.) are excluded.
   *
   * Compatible with the Web API `Request` object (Node.js 18+, Deno, Cloudflare Workers).
   *
   * ```ts
   * app.post("/api/infer", async (req, res) => {
   *   const webReq = new Request(req.url, { method: req.method, headers: req.headers, body: req });
   *   const result = await client.doRequest(webReq);
   *   res.json(result.payload);
   * });
   * ```
   */
  async doRequest(
    request: Request,
    options?: { timeout?: number }
  ): Promise<Result> {
    const headers: Record<string, string> = {};
    request.headers.forEach((value, key) => {
      if (!HOP_BY_HOP.has(key.toLowerCase())) {
        headers[key] = value;
      }
    });

    const payload = await request.text();
    const response = await this.doSync(payload, headers, options);
    if ("result" in response) return response.result;
    return this.waitResult(response.requestId);
  }
}

// ─── LocalClient ─────────────────────────────────────────────────────────────

const DEFAULT_BATCH_SIZE = 10;
const DEFAULT_POLL_INTERVAL_MS = 1_000;
const DEFAULT_WORKERS = 1;

/**
 * LocalClient — poll hub-router for requests, process them, push results back.
 * Fully async, non-blocking. The poll loop runs as an independent Promise.
 *
 * ```ts
 * const client = new LocalClient("https://hub.example.com", "local-key");
 *
 * // Fire-and-forget alongside your server
 * client.run(async (req) => {
 *   const answer = await myModel.infer(req.payload);
 *   return { request_id: req.id, payload: answer, status_code: 200 };
 * });
 * ```
 */
export class LocalClient {
  private readonly baseUrl: string;
  private readonly apiKey: string;
  private readonly batchSize: number;
  private readonly pollIntervalMs: number;
  private readonly workers: number;
  private stopped = false;

  /**
   * @param baseUrl  hub-router base URL, e.g. "https://hub.example.com"
   * @param apiKey   X-Local-API-Key value
   * @param options  Optional overrides
   */
  constructor(
    baseUrl: string,
    apiKey: string,
    options: LocalClientOptions = {}
  ) {
    this.baseUrl = baseUrl.replace(/\/$/, "");
    this.apiKey = apiKey;
    this.batchSize = options.batchSize ?? DEFAULT_BATCH_SIZE;
    this.pollIntervalMs = options.pollInterval ?? DEFAULT_POLL_INTERVAL_MS;
    this.workers = options.workers ?? DEFAULT_WORKERS;
  }

  /**
   * Start the poll loop. Returns a Promise that resolves when {@link stop} is called.
   *
   * Non-blocking — run alongside your server:
   * ```ts
   * client.run(processor);    // no await — runs in background
   * await server.listen(3000);
   * ```
   */
  async run(processor: ProcessorFn): Promise<void> {
    let activeWorkers = 0;
    const waitForWorker = (): Promise<void> =>
      new Promise((resolve) => {
        const check = () => {
          if (activeWorkers < this.workers) resolve();
          else setImmediate(check);
        };
        check();
      });

    while (!this.stopped) {
      let requests: QueuedRequest[];
      try {
        requests = await this.pullBatch();
      } catch {
        await this._sleep(this.pollIntervalMs);
        continue;
      }

      if (requests.length === 0) {
        await this._sleep(this.pollIntervalMs);
        continue;
      }

      for (const req of requests) {
        await waitForWorker();
        if (this.stopped) break;
        activeWorkers++;
        this._processAndPush(processor, req).finally(() => {
          activeWorkers--;
        });
      }
    }

    while (activeWorkers > 0) {
      await this._sleep(50);
    }
  }

  /** Signal the poll loop to stop after the current batch finishes. */
  stop(): void {
    this.stopped = true;
  }

  /** Fetch up to `batchSize` pending requests. */
  async pullBatch(): Promise<QueuedRequest[]> {
    const res = await fetch(
      `${this.baseUrl}/queue/pull?batch=${this.batchSize}`,
      { headers: { "X-Local-API-Key": this.apiKey } }
    );
    if (res.status !== 200) {
      throw new HubRouterError(
        `unexpected status ${res.status}: ${await res.text()}`
      );
    }
    const data = (await res.json()) as PullResponse;
    return data.requests ?? [];
  }

  /** Push a completed result to hub-router. */
  async pushResult(result: Result): Promise<void> {
    const body: Result = {
      ...result,
      completed_at: result.completed_at ?? new Date().toISOString(),
    };
    const res = await fetch(`${this.baseUrl}/queue/result`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "X-Local-API-Key": this.apiKey,
      },
      body: JSON.stringify(body),
    });
    if (res.status === 404) throw new RequestNotFoundError(result.request_id);
    if (res.status !== 204) {
      throw new HubRouterError(
        `unexpected status ${res.status}: ${await res.text()}`
      );
    }
  }

  private async _processAndPush(
    processor: ProcessorFn,
    req: QueuedRequest
  ): Promise<void> {
    let result: Result;
    try {
      result = await processor(req);
    } catch (err) {
      const message = err instanceof Error ? err.message : String(err);
      result = {
        request_id: req.id,
        payload: { error: message },
        status_code: 500,
        error: message,
      };
    }
    try {
      await this.pushResult(result);
    } catch {
      // Best-effort: request may have expired.
    }
  }

  private _sleep(ms: number): Promise<void> {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }
}
